// browserstack-node-sdk/src/bin/load/load/index.js
#!/usr/bin/env node
'use strict';

const {spawn} = require('child_process');
const path = require('path');
const helper = require('../../helpers/helper');
const CLIUtils = require('../v2/cliUtils');
const LoadTestModule = require('../v2/modules/loadTestModule');
const {getLoadConfigPath, sleep} = require('./loadUtils');
const {
  classify,
  messageForExitCode,
  runtimeNotSupportedMessage,
  LT_EXIT_UNSUPPORTED_RUNTIME,
  ERROR_CODES
} = require('./loadErrors');
const {LT_SUPPORTED_NODE} = require('../utils/constants');
const logger = require('../../helpers/logger').winstonLogger;

/**
 * Best-effort downstream failure propagation to the Load Testing solution
 * (SDK-6442): emit an EDS event so a failed load run is visible to the LT
 * backend/support even when the Go binary never started or exited non-zero.
 * Capped via a race so it never blocks the user's exit on a slow/unreachable
 * network. Only meaningful once credentials are resolved (EDS needs auth).
 */
async function reportFailureToLTS(config, errorCode, extra = {}) {
  // EDS needs credentials to authenticate the event. On a pre-credential failure
  // (config parse / missing creds) skip silently rather than fire an
  // unauthenticated call that can only fail and add a noisy warn. This lets every
  // terminal path call reportFailureToLTS uniformly; it delivers wherever creds
  // are resolved and no-ops (quietly) where they are not.
  if (!config || !config.userName || !config.accessKey) {
    logger.debug('Skipping downstream LT failure report: credentials unavailable.');

    return;
  }
  const props = Object.assign({
    feature: 'LTS',
    // event_name becomes the TOP-LEVEL EDS event_name (sendToEDS forwards it),
    // so downstream queries/alerts can actually filter on LoadCliRunFailed rather
    // than the default SDKFeaturePerformance (SDK-6442).
    event_name: 'LoadCliRunFailed',
    errorCode: errorCode || ERROR_CODES.UNKNOWN
  }, extra);
  // Own the rejection on the send promise itself so that if the 3s cap wins the
  // race, a later EDS rejection (the exact failure this telemetry targets) cannot
  // surface as an unhandled rejection during the subsequent sleep/exit.
  const send = Promise.resolve()
    .then(() => helper.sendToEDS(props, config || {}))
    .catch((e) => {
      logger.warn(`Could not report Load Testing failure downstream: ${e && e.message}`);
    });
  let timer;
  const cap = new Promise((resolve) => {
    timer = setTimeout(resolve, 3000);
  });
  await Promise.race([send, cap]);
  clearTimeout(timer);
}

/**
 * Reject unsupported Node runtimes up front (the SDK-6442 / SDK-6188 scenario)
 * with a clear, machine-detectable message, before any download or spawn.
 * Hard-fails below the supported floor; warns (does not block) above the newest
 * validated major so working-but-untested versions are not falsely rejected.
 * Bypass with BROWSERSTACK_SKIP_NODE_CHECK=true.
 */
async function assertSupportedRuntime(config) {
  const skip = String(process.env.BROWSERSTACK_SKIP_NODE_CHECK || '').toLowerCase();
  if (['true', '1', 'yes'].includes(skip)) {
    return;
  }
  const range = LT_SUPPORTED_NODE || {};
  const detectedMajor = parseInt(String(process.versions.node).split('.')[0], 10);
  if (!Number.isFinite(detectedMajor)) {
    return; // can't determine the runtime — don't block
  }
  // Hard-fail ONLY below a confirmed floor. hardMinMajor is intentionally null
  // until the authoritative range is confirmed (SDK-6442), so this branch is
  // dormant and the preflight is warn-only — an unconfirmed bound must not
  // exit(78) a currently-working user.
  if (range.hardMinMajor && detectedMajor < range.hardMinMajor) {
    logger.error(runtimeNotSupportedMessage(detectedMajor, range));
    await reportFailureToLTS(config, ERROR_CODES.UNSUPPORTED_RUNTIME, {detectedNodeMajor: detectedMajor});
    await sleep();
    process.exit(LT_EXIT_UNSUPPORTED_RUNTIME);
  }
  // Outside the validated band → heads-up only, never block.
  if (range.minTestedMajor && detectedMajor < range.minTestedMajor) {
    logger.warn(`Node.js v${detectedMajor} is older than the oldest version validated for BrowserStack Load Testing (Node ${range.minTestedMajor}). It may work; if you hit issues, switch to a validated Node.js version.`);
  } else if (range.maxTestedMajor && detectedMajor > range.maxTestedMajor) {
    logger.warn(`Node.js v${detectedMajor} is newer than the latest version validated for BrowserStack Load Testing (up to Node ${range.maxTestedMajor}). It may work; if you hit issues, switch to a validated Node.js version.`);
  }
}


async function executeLoadTest(yargs) {
  // Hoisted so the outer catch can report it downstream.
  let config = {};
  try {
    const rawArgs = process.argv.slice(2);
    const loadArgs = rawArgs.slice(1);

    logger.debug(`Executing load test with arguments: ${loadArgs}`);

    try {
      logger.debug('Loading configuration using environment variables if available.');
      config = {
        userName: process.env.BROWSERSTACK_USERNAME || process.env.BROWSERSTACK_USER,
        accessKey: process.env.BROWSERSTACK_ACCESS_KEY || process.env.BROWSERSTACK_ACCESSKEY
      };

      if (!config.userName || !config.accessKey) {
        const configPath = getLoadConfigPath(yargs.argv);
        logger.debug(`Configuration path: ${configPath}`);

        if (configPath) {
          config = helper.readConfig(configPath);
          // readConfig returns an {error} sentinel (not a throw) on a YAML
          // parse / read failure. Surface it as a config error rather than
          // letting it fall through to the misleading "Authentication
          // incomplete" message below.
          if (config && config.error) {
            logger.error(`Failed to read Load Testing configuration at ${configPath}: ${config.error}`);
            await reportFailureToLTS(config, ERROR_CODES.CONFIG_PARSE);
            await sleep();
            process.exit(1);
          }
        } else {
          logger.info(`No browserstack-load.yml / browserstack.yml found in ${process.cwd()}; relying on environment variables for credentials.`);
        }
      }
    } catch (error) {
      const {userMessage, code} = classify(error);
      logger.error(`Failed to load Load Testing configuration: ${userMessage}`);
      logger.debug(`Config load error [${code}]: ${error && (error.stack || JSON.stringify(error))}`);
      await reportFailureToLTS(config, code);
      await sleep();  // Added sleep to ensure logs are flushed
      process.exit(1);
    }

    if (!config.userName || !config.accessKey) {
      logger.error('Authentication incomplete. Please add your userName and accessKey to either the browserstack.yml file or as environment variables, then try running the command again.');
      await reportFailureToLTS(config, ERROR_CODES.AUTH_INVALID, {reason: 'missing-credentials'});
      await sleep();
      process.exit(1);
    }

    // Reject unsupported Node runtimes up front, before any download/spawn
    // (the SDK-6442 / SDK-6188 scenario). Credentials are resolved by now, so a
    // rejection can also be reported downstream to the LT backend.
    await assertSupportedRuntime(config);

    // Standalone `load` runs are Load Testing sessions: link the LTS session id
    // (if provided) and emit LTS presence telemetry. Previously only the
    // automation path (browserstackCLI.js) did this.
    try {
      await LoadTestModule.setLTSSessionId(config);
    } catch (ltsErr) {
      logger.debug(`LTS session linking skipped: ${ltsErr && ltsErr.message}`);
    }

    // Get the CLI directory
    const cliDir = CLIUtils.getCliDir();
    if (!cliDir) {
      logger.error('Failed to determine CLI directory');
      await reportFailureToLTS(config, ERROR_CODES.FS_PERMISSION, {reason: 'no-writable-cli-dir'});
      await sleep();
      process.exit(1);
    }

    // Check if CLI binary exists
    let binaryPath = CLIUtils.getExistingCliPath(cliDir);

    // If CLI doesn't exist or needs to be updated, download it
    // Pass config to CLI download/update process
    try {
      if (!binaryPath) {
        logger.debug('CLI binary not found, downloading latest version');
        binaryPath = await CLIUtils.checkAndUpdateCli('', cliDir, config);
      } else {
        logger.debug('CLI binary found, Checking for CLI updates');
        binaryPath = await CLIUtils.checkAndUpdateCli(binaryPath, cliDir, config);
      }
    } catch (error) {
      // Credentials are already confirmed present above (line ~42), so a
      // failure here is NOT a missing-credentials problem. Classify the real
      // cause (network, proxy/SSL, HTTP error, corrupt download, filesystem,
      // disk-full, …) and surface an actionable message instead of the old
      // blanket "Authentication incomplete" that sent valid-credential users
      // to edit browserstack.yml in vain (SDK-6442 / SDK-6188).
      const {code, userMessage} = classify(error);
      logger.error(userMessage);
      logger.debug(`CLI acquisition failed [${code}]: ${error && (error.stack || error.message || JSON.stringify(error))}`);
      await reportFailureToLTS(config, code);
      await sleep();
      process.exit(1);
    }

    if (!binaryPath) {
      logger.error('Could not obtain the BrowserStack Load Testing CLI binary: the download or extraction did not produce a usable binary. Re-run with DEBUG=true for details, or contact BrowserStack support if this persists.');
      logger.debug('checkAndUpdateCli returned an empty binary path');
      await reportFailureToLTS(config, ERROR_CODES.UNKNOWN, {reason: 'empty-binary-path'});
      await sleep();
      process.exit(1);
    }

    logger.debug(`Spawning: ${binaryPath} load ${loadArgs.join(' ')}`);

    // Pass all arguments directly to the CLI binary
    const childProc = spawn(binaryPath, ['load', ...loadArgs], {
      stdio: 'inherit'
    });

    let isExiting = false;

    // Promise that resolves when the child process exits (or fails to spawn).
    const childProcessPromise = new Promise((resolve) => {
      let settled = false;
      let errored = false;
      const settle = (code) => {
        if (settled) {
          return;
        }
        settled = true;
        resolve(code);
      };

      // A failed spawn (binary missing, not executable, wrong arch) emits
      // 'error' — without a listener Node throws an unhandled error and crashes
      // the process with a raw stack trace (the SDK-6188 breakage class).
      // Surface an actionable message and resolve with a deterministic code so
      // the flow exits cleanly instead of hanging or crashing.
      childProc.on('error', (err) => {
        errored = true;
        let code = 1;
        if (err && err.code === 'ENOENT') {
          code = 127;
          logger.error(`Unable to launch the BrowserStack Load Testing CLI binary at ${binaryPath}. The download may be incomplete or corrupt — delete it (under ~/.browserstack/cli) and re-run.`);
        } else if (err && err.code === 'EACCES') {
          code = 126;
          logger.error(`The BrowserStack Load Testing CLI binary at ${binaryPath} is not executable (permission denied). Check file permissions and re-run.`);
        } else {
          const {userMessage} = classify(err);
          logger.error(`Failed to launch the Load Testing CLI: ${userMessage}`);
        }
        logger.debug(`Spawn error details: ${err && (err.stack || err.message)}`);
        settle(code);
      });

      childProc.on('close', (code) => {
        // If the process failed to spawn, 'error' already reported the cause;
        // don't emit a second, less-specific error line here.
        if (!errored) {
          if (code !== 0 && code !== null) {
            // Promote a non-zero exit to an actionable, user-visible error
            // (mirrors the automation flow, which logs non-zero exits at ERROR).
            logger.error(messageForExitCode(code));
          } else {
            logger.debug(`${binaryPath} exited with code ${code}`);
          }
        }
        settle(code === null ? 1 : code);
      });
    });

    // Handle signals like SIGINT (Ctrl+C), SIGTERM, etc.
    ['SIGINT', 'SIGTERM', 'SIGQUIT', 'SIGHUP'].forEach(signal => {
      process.on(signal, () => {
        if (isExiting) {
          return;
        }
        isExiting = true;

        logger.debug(`Received ${signal}, forwarding to child process...`);

        // Forward the signal to the child process
        if (childProc && !childProc.killed) {
          childProc.kill(signal);
          logger.debug('Waiting for child process to exit...');
        }
      });
    });

    // Wait for the child process to complete
    const exitCode = await childProcessPromise;

    // Report a failed run downstream (covers both a failed spawn — settled with
    // 126/127 above — and a non-zero exit from the binary itself). Skip when the
    // non-zero exit was caused by a forwarded signal (Ctrl+C / SIGTERM): that is
    // a user-initiated cancellation, not a failure, and reporting it would emit a
    // spurious LoadCliRunFailed.
    if (exitCode !== 0 && !isExiting) {
      await reportFailureToLTS(config, ERROR_CODES.UNKNOWN, {exitCode});
    } else if (exitCode !== 0 && isExiting) {
      logger.debug(`Load test cancelled by signal (exit code ${exitCode}); not reported as a failure.`);
    }

    // Exit with the same code as the child process
    await sleep();
    process.exit(exitCode);

  } catch (error) {
    const {code, userMessage} = classify(error);
    logger.error(`Failed initiating load test: ${userMessage}`);
    logger.debug(`Unexpected load error [${code}]: ${error && (error.stack || JSON.stringify(error))}`);
    await reportFailureToLTS(config, code);
    await sleep();
    process.exit(1);
  }
}

module.exports = executeLoadTest;
