// browserstack-node-sdk/src/bin/load/load/loadErrors.js
'use strict';

/**
 * Error taxonomy for the Load Testing (LT / LTS) CLI flow — SDK-6442.
 *
 * The standalone `browserstack load` flow historically collapsed every distinct
 * failure (network, proxy/SSL, HTTP error, corrupt download, filesystem,
 * disk-full, config-parse) into a single misleading "Authentication incomplete"
 * message, hiding the real cause at logger.debug. This module is the single
 * source of truth that maps a low-level error to a stable code + an actionable,
 * user-facing message, so the entry, download and exit-code paths all speak the
 * same language instead of ad-hoc strings.
 */

const ERROR_CODES = {
  AUTH_INVALID: 'AUTH_INVALID',
  NETWORK: 'NETWORK',
  SSL: 'SSL',
  DOWNLOAD_HTTP: 'DOWNLOAD_HTTP',
  EXTRACT_CORRUPT: 'EXTRACT_CORRUPT',
  FS_PERMISSION: 'FS_PERMISSION',
  DISK_FULL: 'DISK_FULL',
  CONFIG_PARSE: 'CONFIG_PARSE',
  UPDATE_CHECK_FAILED: 'UPDATE_CHECK_FAILED',
  UNSUPPORTED_RUNTIME: 'UNSUPPORTED_RUNTIME',
  UNKNOWN: 'UNKNOWN'
};

// SDK-side exit code used when the load preflight rejects the runtime, so CI and
// downstream tooling can distinguish "unsupported runtime" from a generic failure.
const LT_EXIT_UNSUPPORTED_RUNTIME = 78;

// Node/libuv error codes grouped by the user-facing class they map to.
const NETWORK_ERROR_CODES = [
  'ECONNREFUSED', 'ECONNRESET', 'ENOTFOUND', 'EAI_AGAIN', 'ETIMEDOUT',
  'ENETUNREACH', 'EHOSTUNREACH', 'EPIPE', 'ECONNABORTED', 'ESOCKETTIMEDOUT'
];
const FS_PERMISSION_CODES = ['EACCES', 'EPERM', 'EROFS'];
// Common OpenSSL/Node TLS failure codes (mirrors helpers/utils SSL_ERROR_CODES intent).
const SSL_ERROR_CODES = [
  'CERT_HAS_EXPIRED', 'DEPTH_ZERO_SELF_SIGNED_CERT', 'SELF_SIGNED_CERT_IN_CHAIN',
  'UNABLE_TO_VERIFY_LEAF_SIGNATURE', 'UNABLE_TO_GET_ISSUER_CERT_LOCALLY',
  'ERR_TLS_CERT_ALTNAME_INVALID', 'CERT_UNTRUSTED'
];
// Specific TLS phrases only — bare 'ssl'/'tls' are too broad and, because SSL is
// classified before network/HTTP, would mislabel any error whose message merely
// contains those substrings. Genuine TLS errors also carry an SSL_ERROR_CODES code.
const SSL_MESSAGE_HINTS = [
  'self signed certificate', 'self-signed certificate', 'unable to verify',
  'unable to get local issuer', 'certificate has expired',
  'ssl routines', 'ssl alert', 'tls handshake', 'wrong version number'
];
// yauzl / zlib signals that the downloaded archive is not a valid/complete zip.
const ZIP_MESSAGE_HINTS = [
  'central directory', 'not a valid zip', 'end of central', 'invalid signature',
  'invalid local file header', 'compressed/uncompressed size', 'invalid comment length'
];
// Dropped the bare 'access key' token: it matched benign text like
// "using access key from environment". The specific 'invalid access key' remains.
const AUTH_MESSAGE_HINTS = [
  'unauthor', 'forbidden', 'access denied', 'invalid username', 'invalid access key',
  'authentication failed', 'invalid credential'
];

/**
 * A typed error for the LT flow. Carries a stable `code` and a user-facing
 * `userMessage` so callers can log the actionable message at logger.error while
 * keeping the raw cause at logger.debug.
 */
class LoadError extends Error {
  constructor(code, userMessage, options = {}) {
    super(userMessage || code);
    this.name = 'LoadError';
    this.code = ERROR_CODES[code] ? code : ERROR_CODES.UNKNOWN;
    this.userMessage = userMessage || defaultMessageFor(this.code, options);
    this.isAuth = this.code === ERROR_CODES.AUTH_INVALID;
    if (options.statusCode) { this.statusCode = options.statusCode }
    if (options.cause !== undefined) { this.cause = options.cause }
  }
}

/** Build a LoadError without `new` at every call site. */
function makeLoadError(code, options = {}) {
  return new LoadError(code, options.userMessage, options);
}

/** Extract a short human-readable cause string from any thrown value. */
function rawCauseString(err) {
  if (err === null || err === undefined) { return '' }
  if (typeof err === 'string') { return err }
  if (err.userMessage || err.message || err.error || err.reason) {
    return err.userMessage || err.message || err.error || err.reason;
  }
  try { return JSON.stringify(err) } catch (_) { return String(err) }
}

/** Default actionable message for a taxonomy code. */
function defaultMessageFor(code, options = {}) {
  const status = options.statusCode ? ` (HTTP ${options.statusCode})` : '';
  switch (code) {
    case ERROR_CODES.AUTH_INVALID:
      return 'Authentication failed: your BrowserStack username/access key was rejected' + status +
        '. Check BROWSERSTACK_USERNAME / BROWSERSTACK_ACCESS_KEY (or the credentials in your browserstack.yml) and try again.';
    case ERROR_CODES.NETWORK:
      return 'Could not reach BrowserStack to set up the Load Testing CLI. Check your network/proxy connectivity and retry. ' +
        'If sdk-assets.browserstack.com is blocked by your firewall, allowlist it.';
    case ERROR_CODES.SSL:
      return 'SSL/TLS error while contacting BrowserStack. If you are behind an SSL-inspecting proxy, set ' +
        'NODE_EXTRA_CA_CERTS (or BROWSERSTACK_EXTRA_CA_CERTS) to your corporate CA bundle and retry.';
    case ERROR_CODES.DOWNLOAD_HTTP:
      return 'Failed to download the Load Testing CLI binary' + status +
        '. Please retry; if this persists, contact BrowserStack support.';
    case ERROR_CODES.EXTRACT_CORRUPT:
      return 'The downloaded Load Testing CLI binary could not be extracted (the archive appears corrupt or incomplete). ' +
        'Delete the cached binary under ~/.browserstack/cli and retry.';
    case ERROR_CODES.FS_PERMISSION:
      return 'Permission denied while writing the Load Testing CLI binary. Ensure the SDK can write to ~/.browserstack ' +
        '(or set BROWSERSTACK_FILES_DIR to a writable directory) and retry.';
    case ERROR_CODES.DISK_FULL:
      return 'Not enough disk space to download the Load Testing CLI binary. Free up disk space and retry.';
    case ERROR_CODES.CONFIG_PARSE:
      return 'Could not read your Load Testing configuration. Check that the browserstack.yml / browserstack-load.yml ' +
        'file exists and is valid YAML.';
    case ERROR_CODES.UPDATE_CHECK_FAILED:
      return 'BrowserStack returned an error while checking for the latest Load Testing CLI' + status +
        '. Please retry; if this persists, contact BrowserStack support.';
    case ERROR_CODES.UNSUPPORTED_RUNTIME:
      return 'Your Node.js runtime is not supported by BrowserStack Load Testing. ' +
        'Switch to a supported Node.js version and re-run. See the BrowserStack Load Testing docs for the supported version matrix.';
    default:
      return 'The Load Testing CLI could not be started. Re-run with DEBUG=true for details, ' +
        'or contact BrowserStack support if this persists.';
  }
}

/**
 * Classify any thrown value into { code, userMessage, isAuth, cause }.
 *
 * Heuristics, in priority order:
 *   1. Already-typed LoadError / objects carrying a known taxonomy `code`.
 *   2. HTTP status (401/403 -> auth, 5xx -> update-check, other 4xx -> download).
 *   3. SSL (TLS error codes or message hints).
 *   4. Filesystem: ENOSPC -> disk-full, EACCES/EPERM/EROFS -> permission.
 *   5. Network (transport error codes).
 *   6. Corrupt archive (yauzl/zlib message hints).
 *   7. Auth-ish API error body messages.
 *   8. Fallback -> UNKNOWN.
 *
 * `nodeRequest` (helpers/helper.js) throws the *parsed response body* for HTTP
 * errors (which loses the status) and the raw error for transport failures, so
 * classify keys off both `err.code` and a (best-effort) preserved `statusCode`.
 */
function classify(err) {
  const cause = rawCauseString(err);

  // 1. Already a typed taxonomy error.
  if (err && typeof err.code === 'string' && ERROR_CODES[err.code]) {
    const userMessage = err.userMessage || defaultMessageFor(err.code, {statusCode: err.statusCode});

    return {code: err.code, userMessage, isAuth: err.code === ERROR_CODES.AUTH_INVALID, cause};
  }

  const nodeCode = err && typeof err.code === 'string' ? err.code : '';
  const status = err && (err.statusCode || (err.response && err.response.statusCode)) || undefined;
  const msg = (err && (err.message || err.error)) ? String(err.message || err.error).toLowerCase() : '';

  const result = (code, statusCode) => ({
    code,
    userMessage: defaultMessageFor(code, {statusCode}),
    isAuth: code === ERROR_CODES.AUTH_INVALID,
    cause
  });

  // 2/3. SSL takes priority over generic status/network because TLS failures can
  // surface with both a code and a transport error.
  if ((nodeCode && SSL_ERROR_CODES.includes(nodeCode)) || SSL_MESSAGE_HINTS.some(h => msg.includes(h))) {
    return result(ERROR_CODES.SSL);
  }
  if (status === 401 || status === 403) { return result(ERROR_CODES.AUTH_INVALID, status) }

  // 4. Filesystem.
  if (nodeCode === 'ENOSPC') { return result(ERROR_CODES.DISK_FULL) }
  if (nodeCode && FS_PERMISSION_CODES.includes(nodeCode)) { return result(ERROR_CODES.FS_PERMISSION) }

  // 5. Network transport.
  if (nodeCode && NETWORK_ERROR_CODES.includes(nodeCode)) { return result(ERROR_CODES.NETWORK) }

  // 6. Corrupt archive.
  if (ZIP_MESSAGE_HINTS.some(h => msg.includes(h))) { return result(ERROR_CODES.EXTRACT_CORRUPT) }

  // 7. Remaining HTTP statuses.
  if (status && status >= 500) { return result(ERROR_CODES.UPDATE_CHECK_FAILED, status) }
  if (status && status >= 400) { return result(ERROR_CODES.DOWNLOAD_HTTP, status) }

  // 8. Auth-ish API body message (when status was stripped by nodeRequest).
  if (AUTH_MESSAGE_HINTS.some(h => msg.includes(h))) { return result(ERROR_CODES.AUTH_INVALID) }

  return result(ERROR_CODES.UNKNOWN);
}

/**
 * Map a non-zero exit code from the Load Testing Go binary to an actionable
 * message.
 *
 * NOTE: the authoritative exit-code contract from the LT Go binary is not yet
 * confirmed (SDK-6442 open question). Keep EXIT_CODE_MESSAGES empty / minimal
 * until backend confirms specific codes; everything falls back to a generic but
 * actionable message. Extend this table — do not invent codes — once confirmed.
 */
const EXIT_CODE_MESSAGES = {
  // Example shape once the binary's contract is confirmed:
  // 3: 'Your Node.js/runtime version is not supported by Load Testing. See <docs>.',
  // 4: 'Invalid Load Testing configuration. Check your browserstack.yml `load:` section.',
};

function messageForExitCode(code) {
  if (Object.prototype.hasOwnProperty.call(EXIT_CODE_MESSAGES, code)) {
    return EXIT_CODE_MESSAGES[code];
  }

  return `Load test failed (exit code ${code}). Re-run with DEBUG=true for details, ` +
    'or contact BrowserStack support if this persists.';
}

/**
 * Build the actionable "Node version not supported" message for the preflight.
 * `detectedMajor` is the running Node major; `range` is constants.LT_SUPPORTED_NODE.
 */
function runtimeNotSupportedMessage(detectedMajor, range = {}) {
  const min = range.hardMinMajor;
  const max = range.maxTestedMajor;
  const supported = (min && max) ? ` (supported: Node ${min}+, validated up to Node ${max})` : '';

  return `Node.js v${detectedMajor} is not supported by BrowserStack Load Testing${supported}. ` +
    'Switch to a supported Node.js version and re-run, or set BROWSERSTACK_SKIP_NODE_CHECK=true to bypass this check.';
}

module.exports = {
  ERROR_CODES,
  LT_EXIT_UNSUPPORTED_RUNTIME,
  LoadError,
  makeLoadError,
  classify,
  defaultMessageFor,
  messageForExitCode,
  runtimeNotSupportedMessage,
  EXIT_CODE_MESSAGES
};
